vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,1019 @@
+
+  clock_t begin;  // for timing performance
+  float time;     // for timing performance
+
+  node* root;
+  LRU* bounder;
+
+  tree()
+  {
+    iter = 0;
+    pass = 0;
+    depth = 0;
+    test_only = false;
+
+    tree_bound = -1;
+    leaf_split = 100;
+
+    scorer_type = 3;
+    router_type = 2;
+
+    begin = clock();
+    time = 0;
+
+    root = nullptr;
+    bounder = nullptr;
+  }
+
+  void deallocate_node(node* n)
+  {
+    if (!n) { return; }
+    for (tree_example* e : n->examples)
+    {
+      VW::free_flatten_example(e->base);
+      VW::free_flatten_example(e->full);
+    }
+    deallocate_node(n->left);
+    deallocate_node(n->right);
+  }
+
+  ~tree()
+  {
+    deallocate_node(root);
+    if (ex) { VW::dealloc_examples(ex, 1); }
+  }
+};
+
+struct rng
+{
+  std::shared_ptr<VW::rand_state> state;
+
+public:
+  rng(std::shared_ptr<VW::rand_state> state) { this->state = std::move(state); }
+
+  using result_type = size_t;
+  static constexpr result_type min() { return 0; }
+  static constexpr result_type max() { return 1; }
+  result_type operator()() { return state->get_and_update_random(); }
+};
+
+node* node_route(node& cn, tree_example& ec)
+{
+  return inner(*ec.base, *cn.router_weights) < cn.router_decision ? cn.left : cn.right;
+}
+
+void tree_init(tree& b)
+{
+  b.iter = 0;
+  b.depth = 0;
+  b.pass = 0;
+
+  b.root = new node();
+
+  b.bounder = new LRU(b.tree_bound);
+
+  // we set this up for repeated use later in the scorer.
+  // we will populate this examples features over and over.
+  b.ex = ::VW::alloc_examples(1);
+  for (int i = 0; i != 1; i++)
+  {
+    b.ex[i].interactions = new std::vector<std::vector<VW::namespace_index>>();
+    b.ex[i].extent_interactions = new std::vector<std::vector<extent_term>>();
+    b.ex[i].indices.push_back(0);
+  }
+}
+
+node& tree_route(tree& b, tree_example& ec)
+{
+  node* cn = b.root;
+  while (cn->internal) { cn = node_route(*cn, ec); }
+  return *cn;
+}
+
+void tree_bound(tree& b, tree_example* ec)
+{
+  tree_example* to_delete = b.bounder->bound(ec);
+
+  if (to_delete == nullptr) { return; }
+
+  node& cn = tree_route(b, *to_delete);
+
+  for (auto iter = cn.examples.begin(); iter != cn.examples.end(); iter++)
+  {
+    if (*iter == to_delete)
+    {
+      cn.examples.erase(iter);
+      return;
+    }
+  }
+}
+
+float scorer_initial(VW::example& ex) { return 1 - std::exp(-std::sqrt(ex.total_sum_feat_sq)); }
+
+/// <summary>
+/// Calculate pairwise difference features for a scorer example
+/// </summary>
+/// <param name="f1">features from scorer example 1</param>
+/// <param name="f2">features from scorer eample 2</param>
+/// <param name="out">the features object calcuated features will be written to</param>
+/// <param name="feature_type"> 1 is absolute difference; 2 is difference</param>
+void scorer_features(features& f1, features& f2, features& out, int feature_type)
+{
+  out.values.clear_noshrink();
+  out.indices.clear_noshrink();
+  out.sum_feat_sq = 0;
+
```

Comment:
Sure. I'm happy to write it that way. The reason noshrink was used at all was because when I was did some performance testing this gave a small increase in performance. This function is one of the most performance critical of the whole tree.